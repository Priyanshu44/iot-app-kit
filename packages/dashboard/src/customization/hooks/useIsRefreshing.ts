import { useEffect, useRef, useState } from 'react';

const useIsRefreshing = (isRefreshing: boolean, delay: number) => {
  const timeout = useRef<number | null | NodeJS.Timeout>(null);
  const [delayLoading, setDelayLoading] = useState(false);

  useEffect(() => {
    clearTimeout(timeout.current as number);
    if (isRefreshing) {
      timeout.current = setTimeout(() => {
        setDelayLoading(isRefreshing);
      }, delay);
    } else {
      setDelayLoading(false);
    }
    return () => {
      clearTimeout(timeout.current as number);
    };
  }, [isRefreshing, delay]);

  return delayLoading;
};

export default useIsRefreshing;
