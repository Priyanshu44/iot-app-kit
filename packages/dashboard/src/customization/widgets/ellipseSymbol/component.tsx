import React from 'react';
import { EllipseWidget } from '../types';
import { SVG_STROKE_DASHED, SVG_STROKE_DOTTED, SVG_STROKE_SOLID } from '../constants';
import { useGridSettings } from '~/components/actions/useGridSettings';

const EllipseWidgetComponent: React.FC<EllipseWidget> = (widget) => {
  const { borderStyle = 'solid', fill = 'none', borderColor = 'black', borderThickness = 5 } = widget.properties;
  const { cellSize } = useGridSettings();

  let strokeString = SVG_STROKE_SOLID;

  if (borderStyle === 'dashed') {
    strokeString = SVG_STROKE_DASHED;
  } else if (borderStyle === 'dotted') {
    strokeString = SVG_STROKE_DOTTED;
  }

  const widthPx = widget.width * cellSize;
  const heightPx = widget.height * cellSize;

  const fitFull = {
    width: widthPx,
    height: heightPx,
  };

  const halfBorderThickness = borderThickness / 2;

  return (
    <div style={fitFull}>
      <svg viewBox={`0 0 ${widthPx} ${heightPx}`} style={fitFull} pointerEvents='none'>
        <ellipse
          cx={widthPx / 2}
          cy={heightPx / 2}
          rx={widthPx / 2 - halfBorderThickness}
          ry={heightPx / 2 - halfBorderThickness}
          stroke={borderColor}
          strokeWidth={borderThickness}
          stroke-dasharray={strokeString}
          fill={fill}
          vectorEffect='non-scaling-stroke'
        />
      </svg>
    </div>
  );
};

export default EllipseWidgetComponent;
