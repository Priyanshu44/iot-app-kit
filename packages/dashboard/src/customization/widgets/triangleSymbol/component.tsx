import React from 'react';
import { TriangleWidget } from '../types';
import { SVG_STROKE_DASHED, SVG_STROKE_DOTTED, SVG_STROKE_SOLID } from '../constants';
import { useGridSettings } from '~/components/actions/useGridSettings';

const TriangleWidgetComponent: React.FC<TriangleWidget> = (widget) => {
  const { borderStyle = 'solid', fill = 'none', borderColor = 'black', borderThickness = 5 } = widget.properties;
  const { cellSize } = useGridSettings();

  let strokeString = SVG_STROKE_SOLID;

  if (borderStyle === 'dashed') {
    strokeString = SVG_STROKE_DASHED;
  } else if (borderStyle === 'dotted') {
    strokeString = SVG_STROKE_DOTTED;
  }

  const widthPx = widget.width * cellSize;
  const heightPx = widget.height * cellSize;

  const fitFull = {
    width: widthPx,
    height: heightPx,
  };

  const halfBorderThickness = borderThickness / 2;

  return (
    <div style={fitFull}>
      <svg
        width={widthPx}
        height={heightPx}
        viewBox={`0 0 ${widthPx} ${heightPx}`}
        style={fitFull}
        pointerEvents='none'
      >
        {/* {we need halfBorderThickness so that the entire svg is visible} */}
        <polygon
          points={`${halfBorderThickness},${heightPx - halfBorderThickness} ${widthPx / 2},${halfBorderThickness} ${
            widthPx - halfBorderThickness
          },${heightPx - borderThickness / 2}`}
          stroke={borderColor}
          strokeWidth={borderThickness}
          stroke-dasharray={strokeString}
          fill={fill}
          vectorEffect='non-scaling-stroke'
        />
      </svg>
    </div>
  );
};

export default TriangleWidgetComponent;
