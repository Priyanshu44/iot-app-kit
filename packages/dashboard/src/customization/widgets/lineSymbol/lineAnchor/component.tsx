import React, { CSSProperties, useEffect, useRef } from 'react';
import './component.css';
import { LineWidget } from '~/customization/widgets/types';
import { XYCoord, useDrag } from 'react-dnd';
import { DashboardWidget } from '~/types';
import { useGridSettings } from '~/components/actions/useGridSettings';
import { useWidgetActions } from '~/customization/hooks/useWidgetActions';

function computeNewPosition(
  initialPosition: XYCoord,
  offsetDifference: XYCoord,
  height: number,
  width: number,
  cellSize: number
) {
  // currently constrained within the bounds of the box
  const { x: initialX, y: initialY } = initialPosition;
  const { x: offsetX, y: offsetY } = offsetDifference;

  const widthPixels = Math.round(width * cellSize);
  const heightPixels = Math.round(height * cellSize);

  const newX = Math.max(0, Math.min(initialX + offsetX, widthPixels));
  const newY = Math.max(0, Math.min(initialY + offsetY, heightPixels));

  return { x: newX, y: newY };
}

const useDragAndUpdate = (widget: LineWidget, anchorType: 'start' | 'end') => {
  const { update: updateWidget } = useWidgetActions<DashboardWidget>();
  // These refs are needed because we want to remember the coordinates at the start of the drag event.
  // We don't want to recalculate the coordinates of the points on every render, only when the drag event completes.
  const initialStartPointRef = useRef({
    x: 0,
    y: 0,
  });
  const initialEndPointRef = useRef({
    x: 0,
    y: 0,
  });

  const { cellSize } = useGridSettings();

  const [{ diff, isDragging }, ref] = useDrag({
    type: 'LineAnchor',
    item: () => {
      initialStartPointRef.current = {
        x: widget.properties.start.x,
        y: widget.properties.start.y,
      };
      initialEndPointRef.current = {
        x: widget.properties.end.x,
        y: widget.properties.end.y,
      };
      return {
        anchorType,
      };
    },
    collect: (monitor) => {
      const offsetDifference = monitor.getDifferenceFromInitialOffset();
      return { diff: offsetDifference, isDragging: monitor.isDragging() };
    },
  });
  useEffect(() => {
    if (isDragging && diff) {
      const initialPoint = anchorType === 'start' ? initialStartPointRef.current : initialEndPointRef.current;

      const { x, y } = computeNewPosition(initialPoint, diff, widget.height, widget.width, cellSize);
      const updatedProperties = {
        ...widget.properties,
        [anchorType]: {
          x,
          y,
        },
      };

      const updatedWidget = {
        ...widget,
        properties: updatedProperties,
      };

      updateWidget(updatedWidget);
    }
  }, [diff?.x, diff?.y]);

  return ref;
};

export const LineAnchor: React.FC<{
  style: CSSProperties;
  anchorType: 'start' | 'end';
  widget: LineWidget;
}> = ({ style, anchorType, widget }) => {
  const dragRef = useDragAndUpdate(widget, anchorType);
  return (
    <div
      id='line-symbol-widget'
      data-testid={'line-' + anchorType + '-anchor'}
      ref={dragRef}
      style={style}
      className='line-anchor'
    />
  );
};
