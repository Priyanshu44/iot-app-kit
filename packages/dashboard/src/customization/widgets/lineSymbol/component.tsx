import React, { CSSProperties } from 'react';
import './component.css';
import { LineWidget } from '~/customization/widgets/types';
import { SVG_STROKE_DASHED, SVG_STROKE_DOTTED, SVG_STROKE_SOLID } from '../constants';
import { LineAnchor } from './lineAnchor/component';
import { useIsSelected } from '~/customization/hooks/useIsSelected';

const LineWidgetComponent: React.FC<LineWidget> = (widget) => {
  const isSelected = useIsSelected(widget);

  const fitFull = {
    width: '100%',
    height: '100%',
  };

  const {
    lineStyle = 'solid',
    color = 'black',
    thickness = 5,
    start = {
      x: 25,
      y: 200,
    },
    end = {
      x: 375,
      y: 200,
    },
  } = widget.properties;

  let strokeString = SVG_STROKE_SOLID;

  if (lineStyle === 'dashed') {
    strokeString = SVG_STROKE_DASHED;
  } else if (lineStyle === 'dotted') {
    strokeString = SVG_STROKE_DOTTED;
  }

  const startAnchorStyle: CSSProperties = {
    position: 'absolute',
    left: `${start.x}px`,
    top: `${start.y}px`,
  };

  const endAnchorStyle: CSSProperties = {
    position: 'absolute',
    left: `${end.x}px`,
    top: `${end.y}px`,
  };

  return (
    <div className='fit-full'>
      {isSelected && <LineAnchor style={startAnchorStyle} anchorType='start' widget={widget} />}
      <svg style={fitFull} pointerEvents='none'>
        <line
          x1={start.x}
          y1={start.y}
          x2={end.x}
          y2={end.y}
          stroke={color}
          strokeWidth={thickness}
          stroke-dasharray={strokeString}
        />
      </svg>
      {isSelected && <LineAnchor style={endAnchorStyle} anchorType='end' widget={widget} />}
    </div>
  );
};

export default LineWidgetComponent;
