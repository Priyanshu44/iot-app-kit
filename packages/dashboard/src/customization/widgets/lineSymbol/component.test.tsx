import React from 'react';
import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { useDragAndUpdate } from './lineAnchor/component'; // Correct path to your hook
import { DndProvider, useDrag } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import LineWidgetComponent from './component';
import { Provider } from 'react-redux';
import { configureDashboardStore } from '~/store';
import { useIsSelected } from '~/customization/hooks/useIsSelected';
import { DashboardState } from '~/store/state';
import { MOCK_LINE_WIDGET } from '~/../testing/mocks';
import { useWidgetActions } from '~/customization/hooks/useWidgetActions';

jest.mock('~/customization/hooks/useIsSelected', () => ({
  useIsSelected: jest.fn(),
}));

const initialState: Partial<DashboardState> = {
  dashboardConfiguration: {
    widgets: [MOCK_LINE_WIDGET],
    viewport: { duration: '5m' },
  },
  selectedWidgets: [MOCK_LINE_WIDGET],
};

describe('LineWidgetComponent', () => {
  let mockStore;

  beforeEach(() => {
    // jest.resetAllMocks();
    mockStore = configureDashboardStore(initialState);
  });

  const renderComponentWithDnd = (properties) => {
    return render(
      <Provider store={mockStore}>
        <DndProvider backend={HTML5Backend}>
          <LineWidgetComponent
            id='line-id'
            x={1}
            y={2}
            z={3}
            height={100}
            width={100}
            type='line-widget'
            properties={properties}
          />
        </DndProvider>
      </Provider>
    );
  };

  describe('Rendering', () => {
    beforeEach(() => {
      const { container } = renderComponentWithDnd({
        lineStyle: 'solid',
        color: 'black',
        thickness: 5,
      });
      console.log(container.innerHTML);
    });

    [
      { lineStyle: 'solid', color: 'black', thickness: 5, isSelected: true },
      { lineStyle: 'dashed', color: 'black', thickness: 5, isSelected: true },
      { lineStyle: 'dotted', color: 'black', thickness: 5, isSelected: true },
      { lineStyle: 'solid', color: 'red', thickness: 5, isSelected: false },
      { lineStyle: 'dashed', color: 'blue', thickness: 10, isSelected: false },
      { lineStyle: 'dotted', color: 'green', thickness: 15, isSelected: false },
    ].forEach((configuration) => {
      const { lineStyle, color, thickness, isSelected } = configuration;
      const properties = { lineStyle, color, thickness };
      (useIsSelected as jest.Mock).mockImplementation(() => isSelected);
      it(`should render with ${JSON.stringify(properties)} correctly`, () => {
        const { container } = renderComponentWithDnd(properties);
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('useDragAndUpdate', () => {
    const mockWidget = {
      height: 200,
      width: 200,
      properties: {
        start: {
          x: 50,
          y: 50,
        },
        end: {
          x: 150,
          y: 150,
        },
      },
    };

    // Similarly for 'end' anchor and other test scenarios
  });
});
