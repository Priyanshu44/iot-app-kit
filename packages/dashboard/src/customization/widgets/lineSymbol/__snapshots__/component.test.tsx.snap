// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"dashed","color":"black","thickness":5} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="black"
        stroke-dasharray="10,5"
        stroke-width="5"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"dashed","color":"blue","thickness":10} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="blue"
        stroke-dasharray="10,5"
        stroke-width="10"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"dotted","color":"black","thickness":5} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="black"
        stroke-dasharray="3,3"
        stroke-width="5"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"dotted","color":"green","thickness":15} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="green"
        stroke-dasharray="3,3"
        stroke-width="15"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"solid","color":"black","thickness":5} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="black"
        stroke-dasharray="1,0"
        stroke-width="5"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;

exports[`LineWidgetComponent Rendering should render with {"lineStyle":"solid","color":"red","thickness":5} correctly 1`] = `
<div>
  <div
    class="fit-full"
  >
    <svg
      pointer-events="none"
      style="width: 100%; height: 100%;"
    >
      <line
        stroke="red"
        stroke-dasharray="1,0"
        stroke-width="5"
        x1="25"
        x2="375"
        y1="200"
        y2="200"
      />
    </svg>
  </div>
</div>
`;
